tests/test_unittest/test_gtc/test_gtir.py

```diff
@@ -91,43 +96,171 @@ def test_naive_python_copy(copy_computation):
-                                            left=FieldAccess.centered(name="a"),
-                                            right=BinaryOp(
-                                                left=FieldAccess(
-                                                    name="b",
-                                                    offset=CartesianOffset(i=-1, j=0, k=0),
-                                                ),
-                                                right=FieldAccess(
-                                                    name="b", offset=CartesianOffset(i=1, j=0, k=0)
-                                                ),
-                                                op=BinaryOperator.ADD,
-                                            ),
-                                        )
-                                    )
-                                ],
+        vertical_loops=[
+            VerticalLoop(
+                loop_order=LoopOrder.FORWARD,
+                vertical_intervals=[
+                    VerticalInterval(
+                        start=AxisBound(level=LevelMarker.START, offset=0),
+                        end=AxisBound(level=LevelMarker.END, offset=0),
+                        body=[
+                            ParAssignStmt(
+                                left=FieldAccess.centered(name="a"),
+                                right=BinaryOp(
+                                    left=FieldAccess(
+                                        name="b",
+                                        offset=CartesianOffset(i=-1, j=0, k=0),
+                                    ),
+                                    right=FieldAccess(
+                                        name="b", offset=CartesianOffset(i=1, j=0, k=0)
+                                    ),
+                                    op=ArithmeticOperator.ADD,
+                                ),
                             )
                         ],
                     )
-                ]
+                ],
             )
         ],
     )
     assert ast.parse(PythonNaiveCodegen.apply(horizontal_avg))
+
+
+def test_ExprBaseclassIsNotInstantiatable():
+    with pytest.raises(TypeError):
+        Expr()
+
+
+def test_StmtBaseclassIsNotInstantiatable():
+    with pytest.raises(TypeError):
+        Stmt()
+
+
+def test_DeclBaseclassIsNotInstantiatable():
+    with pytest.raises(TypeError):
+        Decl()
+
+
+# Validation tests
+def test_ParAssignStmtWithVerticalOffsetIsOk():
+    ParAssignStmt(
+        left=FieldAccessBuilder("foo").offset(CartesianOffset(i=0, j=0, k=1))(), right=DummyExpr()
+    )
+
+
+def test_ParAssignStmtWithHorizontalOffsetIsError():
+    with pytest.raises(ValidationError):
+        ParAssignStmt(
+            left=FieldAccessBuilder("foo").offset(CartesianOffset(i=1, j=0, k=0))(),
+            right=DummyExpr(),
+        )
+
+
+arithmetic_type = DataType.FLOAT32
+another_arithmetic_type = DataType.INT32
```

Comment:
What do you think about this?
I want to express: I don't care what arithmetic type it is, but I care that there is another one which is not the same.

Reply:
Extensive way of doing this:

```python
import itertools

ARITHMETIC_TYPES = set(DataType) - {DataType.AUTO, DataType.INVALID, DataType.DEFAULT}

def two_different_types():
    for type_a, type_b in list(itertools.product(ARITHMETIC_TYPES, ARITHMETIC_TYPES)):
        if not type_a == type_b:
            yield type_a, type_b

@pytest.fixture(params=list(two_different_types()))
def two_types(request):
    yield request.param
```

Reply:
Then every test that uses the `two_types` fixture will run for all combinations of different arithmetic types.

Reply:
I thought about it. I think this pattern makes sense only if we have functionality where behavior differs depending on DataType. I think introducing it now would just distract from the code under test.